File: ros/src/ebug_cpp/src/byte_rectifier.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"

#include <sensor_msgs/msg/image.hpp>
#include <sensor_msgs/msg/compressed_image.hpp>

using namespace std::chrono_literals;

#include <rclcpp_components/register_node_macro.hpp>
RCLCPP_COMPONENTS_REGISTER_NODE(ebug::ByteRectifier)


namespace ebug
{
    class ByteRectifier : public rclcpp::Node
    {
        private:
            rclcpp::Publisher<sensor_msgs::msg::CompressedImage>::SharedPtr m_CompressedImagePublisher;
            rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr m_ImageSubscription;

        public:
            ByteRectifier(const rclcpp::NodeOptions& options) : Node("ByteRectifier", options)
            {
                m_CompressedImagePublisher = this->create_publisher<sensor_msgs::msg::CompressedImage>("image_compressed", 10);

                m_ImageSubscription = this->create_subscription<sensor_msgs::msg::Image>("image_raw", 10, std::bind(&ByteRectifier::camera_callback, this, std::placeholders::_1));
            }

        private:
            void camera_callback(const sensor_msgs::msg::Image::ConstSharedPtr& image) const
            {
                sensor_msgs::msg::CompressedImage cimage;

                cimage.header = image->header;
                cimage.data = image->data;

                cimage.format = "jpeg";
                
                m_CompressedImagePublisher->publish(cimage);
            }
    };
}

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ebug::ByteRectifier>());
  rclcpp::shutdown();
  return 0;
}

File: ros/src/ebug_client/ebug_client/ByteRectifier.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile

from sensor_msgs.msg import Image, CompressedImage

class ByteRectifier(Node):
    def __init__(self):
        super().__init__(self.__class__.__name__)

    
        qos_profile = QoSProfile(depth=10)
        self.sub = self.create_subscription(Image, "image_raw", self.image_callback, qos_profile)
        self.pub = self.create_publisher(CompressedImage, "image_compressed", qos_profile)


    def image_callback(self, image: Image):
        cimage = CompressedImage()

        cimage.header = image.header
        cimage.data = image.data
        cimage.format = 'jpeg'

        self.pub.publish(cimage)




## Boilerplate

def main():
    rclpy.init()
    node = ByteRectifier()
    
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        return
    finally:
        node.destroy_node()

    rclpy.shutdown()

if __name__ == '__main__':
    main()